src/error.rs

```diff
@@ -1,9 +1,22 @@
+use crate::collection::Id;
 use thiserror::Error;
 
-#[derive(Error, Debug)]
+#[derive(Error)]
 /// Typed error for `typed_index_collection`.
-pub enum Error {
+pub enum Error<T: Id<T>> {
     /// This error occurs when an identifier already exists.
-    #[error("identifier {0} already exists")]
-    IdentifierAlreadyExists(String),
+    #[error("identifier {} already exists", .0.id())]
+    IdentifierAlreadyExists(T),
+}
+
+impl<T: Id<T>> std::fmt::Debug for Error<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::IdentifierAlreadyExists(obj) => f
+                .debug_struct("IdentifierAlreadyExists Error")
+                .field("id", &obj.id())
+                .field("type", &std::any::type_name::<T>().to_string())
```

Comment:
Actually, you need to return something that is `&dyn Debug` and indeed, `&str` doesn't make `str` match with `dyn Debug`. However, `&str` is `dyn Debug`. So you just had to give a reference on the result of `type_name()` which would be of type `&&str`.
```suggestion
                .field("type", &std::any::type_name::<T>())
```

Reply:
Indeed… fixed and rebased